### ISSUE TITLE
Some logs/feedback from weak hardware conf

### ISSUE THREAD

**Issue by HumanG33k** [iv-org/invidious#721]
For try invidious i just pick up an old laptop and install on it.
This is a part of the configuration : 
OS: Debian GNU/Linux 10 (buster) i686
Kernel: 4.19.0-5-686-pae
CPU: Intel Pentium M 1.73GHz (1) @ 1.733GHz
GPU: Intel Mobile 915GM/GMS/910GML Express Graphics Controller
Memory: 1331MiB / 1943MiB

Pretty old, pretty weak.

So after near than 12 hours of work, this is terminal log.
Some of the line are truncate because i resize the xterm window.

I down know if it can be usefull or improve except by hardware upgrade.

also because of cpu/mem usage videos are stop playing with the following error : 
The media playback was aborted due to a corruption problem or because the media used features your browser did not support.
 
```bash 
./invidious -o invidious.log
Unhandled exception in spawn: Closed stream (IO::Error)
GC Warning: Repeated allocation of very large block (appr. size 
        May lead to memory leak and poor performance
  from /usr/share/crystal/src/io.cr:0:5 in 'write'
  from /usr/share/crystal/src/http/server/response.cr:80:7 in 'w
  from /usr/share/crystal/src/io.cr:481:7 in 'write_utf8'
  from /usr/share/crystal/src/string.cr:4113:5 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in 'run'
  from ???
GC Warning: Repeated allocation of very large block (appr. size 2101248):
        May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 2101248):
        May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 2101248):
        May lead to memory leak and poor performance

```

**Comment by Perflyst** [comment]
I also have these kind of errors in syslog
```
Aug 18 16:46:12 invidious invidious[22521]: GC Warning: Repeated allocation of very large block (appr. size 4198400):
Aug 18 16:46:12 invidious invidious[22521]: #011May lead to memory leak and poor performance
Aug 18 16:46:39 invidious invidious[22521]: GC Warning: Repeated allocation of very large block (appr. size 4198400):
Aug 18 16:46:39 invidious invidious[22521]: #011May lead to memory leak and poor performance
Aug 18 16:46:57 invidious invidious[22521]: GC Warning: Repeated allocation of very large block (appr. size 4198400):
Aug 18 16:46:57 invidious invidious[22521]: #011May lead to memory leak and poor performance
Aug 18 16:47:01 invidious CRON[3908]: (root) CMD (curl -o /dev/null -s -m 10 http://localhost:3000 || systemctl restart invidious | echo $(date '+%Y-%m-%d %H:%M:%S') "+00:00 Restarted Invidious via cron" >> /root/invidious-cron.log)
```

As you can see the restart cronjob kicks in which means invidious is either extremly slow or not working anymore at all. This happens multiple times. After it ends, I get `Too many open files (Errno)` for ~50 seconds and after that again

```
Aug 18 17:06:52 invidious invidious[22521]: GC Warning: Repeated allocation of very large block (appr. size 3444736):
Aug 18 17:06:52 invidious invidious[22521]: #011May lead to memory leak and poor performance
Aug 18 17:07:01 invidious CRON[4134]: (root) CMD (curl -o /dev/null -s -m 10 http://localhost:3000 || systemctl restart invidious | echo $(date '+%Y-%m-%d %H:%M:%S') "+00:00 Restarted Invidious via cron" >> /root/invidious-cron.log)
```

For around 2 minutes, than no error logs but cronjob restarts followed by
```
Aug 18 18:47:34 invidious invidious[22521]: Unhandled exception in spawn: Missing hash key: "videoDetails" (KeyError)
Aug 18 18:47:34 invidious invidious[22521]:   from /usr/share/crystal/src/int.cr:190:13 in '[]'
Aug 18 18:47:34 invidious invidious[22521]:   from src/invidious/videos.cr:1177:11 in 'fetch_video'
Aug 18 18:47:34 invidious invidious[22521]:   from src/invidious/videos.cr:904:17 in 'get_video'
Aug 18 18:47:34 invidious invidious[22521]:   from src/invidious/videos.cr:895:1 in '->'
Aug 18 18:47:34 invidious invidious[22521]:   from /usr/share/crystal/src/fiber.cr:255:3 in 'run'
Aug 18 18:47:34 invidious invidious[22521]:   from ???
```

Followed by

```
Aug 19 00:32:52 invidious kernel: [5134327.569023] memory: usage 3532812kB, limit 4194304kB, failcnt 84355
Aug 19 00:32:52 invidious kernel: [5134327.569121] memory+swap: usage 4718592kB, limit 4718592kB, failcnt 1455184
Aug 19 00:32:52 invidious kernel: [5134327.569212] kmem: usage 89276kB, limit 9007199254740988kB, failcnt 0
Aug 19 00:32:52 invidious kernel: [5134327.569307] Memory cgroup stats for /lxc/806: cache:0KB rss:0KB rss_huge:0KB shmem:0KB mapped_file:0KB dirty:0KB writeback:0KB swap:0KB inac
tive_anon:0KB active_anon:0KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug 19 00:32:52 invidious kernel: [5134327.569456] Memory cgroup stats for /lxc/806/ns: cache:236KB rss:2560KB rss_huge:0KB shmem:236KB mapped_file:0KB dirty:0KB writeback:0KB swa
p:127516KB inactive_anon:1664KB active_anon:1132KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug 19 00:32:52 invidious kernel: [5134327.569621] Memory cgroup stats for /lxc/806/ns/user.slice: cache:0KB rss:0KB rss_huge:0KB shmem:0KB mapped_file:0KB dirty:0KB writeback:0KB
 swap:0KB inactive_anon:0KB active_anon:0KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug 19 00:32:52 invidious kernel: [5134327.569776] Memory cgroup stats for /lxc/806/ns/user.slice/user-0.slice: cache:0KB rss:0KB rss_huge:0KB shmem:0KB mapped_file:0KB dirty:0KB
writeback:0KB swap:0KB inactive_anon:0KB active_anon:0KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug 19 00:32:52 invidious kernel: [5134327.569940] Memory cgroup stats for /lxc/806/ns/user.slice/user-0.slice/user@0.service: cache:0KB rss:364KB rss_huge:0KB shmem:0KB mapped_fi
le:0KB dirty:0KB writeback:0KB swap:672KB inactive_anon:188KB active_anon:176KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug 19 00:32:52 invidious kernel: [5134327.570105] Memory cgroup stats for /lxc/806/ns/user.slice/user-0.slice/session-257.scope: cache:8KB rss:10344KB rss_huge:0KB shmem:8KB mapp
ed_file:0KB dirty:0KB writeback:0KB swap:3544KB inactive_anon:5176KB active_anon:5176KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug 19 00:32:52 invidious kernel: [5134327.570276] Memory cgroup stats for /lxc/806/ns/init.scope: cache:8KB rss:880KB rss_huge:0KB shmem:8KB mapped_file:0KB dirty:0KB writeback:0
KB swap:8KB inactive_anon:468KB active_anon:420KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug 19 00:32:52 invidious systemd[1]: invidious.service: Main process exited, code=killed, status=9/KILL
Aug 19 00:32:52 invidious systemd[1]: invidious.service: Unit entered failed state.
Aug 19 00:32:52 invidious systemd[1]: invidious.service: Failed with result 'signal'.
Aug 19 00:33:01 invidious CRON[9226]: (root) CMD (curl -o /dev/null -s -m 10 http://localhost:3000 || systemctl restart invidious | echo $(date '+%Y-%m-%d %H:%M:%S') "+00:00 Resta
rted Invidious via cron" >> /root/invidious-cron.log)
Aug 19 00:33:07 invidious systemd[1]: invidious.service: Service hold-off time over, scheduling restart.
Aug 19 00:33:07 invidious systemd[1]: Stopped Invidious (An alternative YouTube front-end).
Aug 19 00:33:07 invidious systemd[1]: invidious.service: Failed to reset devices.list: Operation not permitted
Aug 19 00:33:07 invidious systemd[1]: Started Invidious (An alternative YouTube front-end).
Aug 19 00:33:11 invidious systemd[1]: Stopping Invidious (An alternative YouTube front-end)...
Aug 19 00:33:11 invidious systemd[1]: Stopped Invidious (An alternative YouTube front-end).
Aug 19 00:33:11 invidious systemd[1]: invidious.service: Failed to reset devices.list: Operation not permitted
Aug 19 00:33:11 invidious systemd[1]: Started Invidious (An alternative YouTube front-end).
```

And some more cron restarts which end after a while.

**Comment by elypter** [comment]
i have a T2080, 945G, 2gb and invidio.us works fine with firefox and mint 19.2 x32. what browser do you use?

**Comment by Perflyst** [comment]
This has nothing to do with invidio.us @elypter 
It is about running invidious yourself, not using invidio.us

**Comment by HumanG33k** [comment]
This time a crash occurs. I don’t know if it was kill by the kernel or just did it. 

```bash
GC Warning: Repeated allocation of very large block (appr. size 2101248):
        May lead to memory leak and poor performance
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
GC Warning: Failed to expand heap by 16777216 bytes
Unhandled exception in spawn: Cannot allocate new fiber stack: Cannot allocate memory (Errno)
  from /usr/share/crystal/src/fiber/stack_pool.cr:46:7 in 'checkout'
  from /usr/share/crystal/src/fiber.cr:45:5 in 'new'
  from /usr/share/crystal/src/concurrent.cr:63:3 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in 'run'
  from ???
Unhandled exception in spawn: Cannot allocate new fiber stack: Cannot allocate memory (Errno)
  from /usr/share/crystal/src/fiber/stack_pool.cr:46:7 in 'checkout'
  from /usr/share/crystal/src/fiber.cr:45:5 in 'new'
  from /usr/share/crystal/src/concurrent.cr:63:3 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in 'run'
  from ???
GC Warning: Repeated allocation of very large block (appr. size 2101248):
        May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 2101248):
        May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 2101248):
        May lead to memory leak and poor performance

```

**Comment by Perflyst** [comment]
Well, this looks like OOM. Check the syslog or dmesg, I think your operating system has just killed it

**Comment by HumanG33k** [comment]
Hello,

i don’t find anything in logs, but next time i will try to see as soon as it "crash".

A new log after my last crash : 
```bash
Invalid memory access (signal 11) at address 0x10
[0x928fa3] ???
[0x4f2613] __crystal_sigfault_handler +211
[0x93e78f] sigfault_handler +31
[0xb7ed6d88] ???
[0x93bc9a] GC_allochblk_nth +202
[0x93c0c1] GC_allochblk +177
[0x92e597] GC_alloc_large +119
[0x92eacd] GC_generic_malloc +333
[0x92ece4] GC_malloc_kind_global +228
[0x93a2c5] GC_malloc_kind +213
[0x92ed8a] GC_malloc +26
[0x4f24fc] ???
[0xb7d276a2] xmlHashCreate +82
[0xb7d74216] xmlXPathNewContext +134
[0x5fb1a5] ???
[0x607ca4] ???
[0x607397] ???
[0x5ea9e3] ???
[0x5e9958] ???
[0x4edef6] ???
[0x0] ???
```

**Comment by Perflyst** [comment]
Please take a look into /var/log/syslog and syslog.1 syslog.2.gz when the error occured

Also enter "dmesg -T -l err" and "dmesg -T -l warn" if you find anything about out of memory killer

**Comment by HumanG33k** [comment]
Nothing about memory or invidious.

**Comment by omarroth** [comment]
Would you mind building from `api-only` and comparing performance and to see if you encounter the same issue with the GC?

**Comment by Perflyst** [comment]
So I got a few OOM kills
```
[557211.462236] Out of memory: Kill process 19938 (invidious) score 866 or sacrifice child
[557211.462280] Killed process 19938 (invidious) total-vm:43834264kB, anon-rss:3477012kB, file-rss:60kB, shmem-rss:0kB
[630782.678432] Out of memory: Kill process 13483 (invidious) score 853 or sacrifice child
[630782.678475] Killed process 13483 (invidious) total-vm:50652020kB, anon-rss:3403124kB, file-rss:3876kB, shmem-rss:0kB
```

This seems to be totally random and the machine has more than 10GB of ram.

**Comment by HumanG33k** [comment]
@omarroth how can do that ?

**Comment by omarroth** [comment]
```bash
$ git checkout api-only
$ crystal build src/invidious.cr
```

I've also made a couple of changes to master that should improve memory usage if you'd like to update and build again.

**Comment by HumanG33k** [comment]
i will test that, i have to migrate to master, i use stable currently. i come back when it s done.

**Comment by TheFrenchGhosty** [comment]
Closed in favor of #1438
